File: src/algorithms/strings/palindrome/palindrome-check/sources/palindrome-check.py
# Palindrome Check — Two-pointer approach
# Returns True if the string reads the same forwards and backwards.
# Time: O(n), Space: O(1)


def palindrome_check(text: str) -> bool:
    left_index = 0  # @step:initialize
    right_index = len(text) - 1  # @step:initialize

    while left_index < right_index:  # @step:compare
        if text[left_index] != text[right_index]:  # @step:compare
            return False  # @step:mismatch
        left_index += 1  # @step:match
        right_index -= 1  # @step:match

    return True  # @step:complete

File: src/algorithms/strings/palindrome/palindrome-check/sources/palindrome-check.go
// Palindrome Check — Two-pointer approach
// Returns true if the string reads the same forwards and backwards.
// Time: O(n), Space: O(1)

package main

func palindromeCheck(text string) bool {
	chars := []rune(text)
	leftIndex := 0 // @step:initialize
	rightIndex := len(chars) - 1 // @step:initialize

	for leftIndex < rightIndex {
		// @step:compare
		if chars[leftIndex] != chars[rightIndex] {
			return false // @step:mismatch
		}
		leftIndex++ // @step:match
		rightIndex-- // @step:match
	}

	return true // @step:complete
}
